### ISSUE TITLE
Writelog splitstore

### ISSUE THREAD

**Issue by cryptoquick** [cberner/redb#517]
I was thinking, maybe one way to reduce the impact on CoW fs and SSDs like in #509 is to split out the values from the keys. The keys could use redb as it is now (the API I like very much btw, schemas are really cool), but every value is a pointer to an offset and length within a separate, append-only writelog. This log would be written in 4KB chunks, to better align to filesystem sectors.

Before: Key maps to value, both in a BTree
After: Two files; Index where key maps to offset and length kept separately within log opened in append+read mode.

For background, append-only logs are better on SSDs than BTrees because they have to balance occasionally; usually this is rectified by using an LSM tree approach, but that also has its own drawbacks.

Some advantages/disadvantages of a splitstore approach:

Advantages: It might dramatically improve write performance, and also reduce wear and tear on SSDs. This would also be a better choice when batched writes are regularly at least 4KB in size, and when the size of values regularly exceeds 12 bytes.

Disadvantages: It would take up more space overall on the drive, and old values would need to be retained; updates would simply point to the offset for the new value. 

The way I'd imagine it being implemented in practice is gated behind an optional, non-default `splitstore` feature flag for those who know their db will exceed the multiple gigabytes. Or maybe it wouldn't be gated behind a feature, but instead added as a builder argument, say, `open_writelog`.

As for the index offset and length values, I would imagine reasonable defaults would be `(u64, u32)`; 8 bytes for a byte offset into the writelog, and then each value could be a maximum of 4GB in size, assuming there wouldn't be a circumstance where a user would want to write more than 4GB in a single value. Conversely, for more optimization, one could keep separate writelog files, 0-255, each file could be up to 4GB in size, and then each value could be limited to 64KB, which is also reasonable, depending on the use-case. This would result in index values being (u8, u32, u16), or, 7 bytes instead of 12, at the increased complexity of multiple open files and needing to configure them in a directory instead of a single file. This might be over-optimization, but since this would make every value kept in the database a fixed size, every byte counts.

Oh, also, one more thing, for values that are smaller than what would make sense to store in the writelog, maybe the value itself could be written instead. I don't think an enum would be needed, since the only thing needed to know it's the actual value is if it's under 12 bytes. Then, if the value ever exceeds 12 bytes, further updates would be written to the writelog.

**Comment by cberner** [comment]
That's an interesting idea. It sounds like you can implement it in a separate library, and I think that's better than putting this into redb. For example, you can store the `(offset, length)` as the value of a redb `Table` and then write the data into a separate file

**Comment by casey** [comment]
`ord` has a lot of weird append-only writes. Read-only fields with incrementing keys, for example. I've actually been thinking about doing this for certain things, for example if we were to actually save transaction data in redb:

> For example, you can store the `(offset, length)` as the value of a redb `Table` and then write the data into a separate file

Transaction data is saved and never changes, so a flat file would be ideal. Ovbs might not make sense for redb, but if it were a common workload it might.

**Comment by cryptoquick** [comment]
@cberner I can see that, but it might be more challenging to maintain the table schema interface that makes this library so nice to use.

Also, about memory mapping, today I learned the 64 bit OSes allocate 48 bits to memory mapping, so if memory mapping was also used for the writelog, value overhead would actually be 10 bytes.

**Comment by cberner** [comment]
> the table schema interface that makes this library so nice to use

Thanks! Ya I really like it too.

I added an example showing how to do this. It seems straight forward to preserve the table schema, but feel free to re-open or open a new issue if you run into problems implementing something like that.
